core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,731 @@
+	file_len = encrypted_data_len - file_header_len;
+
+	/* allocate buffer to hold encrypted file context + header */
+	encrypted_data = malloc(encrypted_data_len);
+	if (!encrypted_data) {
+		res = -1;
+		goto exit_free_fd;
+	}
+
+	/*
+	 * init internal status, also allocate buffer for
+	 * decrypted file content
+	 */
+	fd->nw_fd = nw_fd;
+	fd->flags = flags;
+	fd->pos = 0;
+	fd->len = file_len;
+	fd->data = malloc(file_len);
+	if (!fd->data) {
+		res = -1;
+		goto exit_free_encrypted_data;
+	}
+
+	/* read encrypted file content */
+	res = read_file(nw_fd, encrypted_data, encrypted_data_len);
+	if (res != (int)encrypted_data_len) {
+		res = -1;
+		goto exit_free_decrypted_data;
+	}
+	DMSG("%d bytes read", res);
+
+	/* decrypt and authenticate the file content */
+	ret = tee_enc_fs_file_decryption(encrypted_data, encrypted_data_len,
+			fd->data, &file_len);
+	if (ret != TEE_SUCCESS) {
+		res = -1;
+		goto exit_free_decrypted_data;
+	}
+
+	res = handle_get(&fs_handle_db, fd);
+
+exit_free_decrypted_data:
+	if (res < 0)
+		free(fd->data);
+exit_free_encrypted_data:
+	free(encrypted_data);
+exit_free_fd:
+	if (res < 0)
+		free(fd);
+exit:
+	return res;
+}
+
+static int tee_fs_close(int fd)
+{
+	int res = -1;
+	size_t encrypted_data_len;
+	size_t file_header_len = tee_enc_fs_get_file_header_size();
+	void *encrypted_data;
+	TEE_Result ret;
+	struct tee_fs_fd *fdp = handle_put(&fs_handle_db, fd);
+	struct tee_fs_rpc head = { 0 };
+
+	DMSG("close, fd=%d", fd);
+
+	if (!fdp)
+		return -1;
+
+	encrypted_data_len = fdp->len + file_header_len;
+
+	/* allocate buffer to hold encrypted file content */
+	encrypted_data = malloc(encrypted_data_len);
+	if (!encrypted_data)
+		goto exit_free_fd;
+
+	/* encrypt the file content and calculate MAC */
+	ret = tee_enc_fs_file_encryption(fdp->data, fdp->len,
+			encrypted_data, &encrypted_data_len);
+	if (ret != TEE_SUCCESS)
+		goto exit_free_encrypted_data;
+
+	res = lseek(fdp->nw_fd, 0, TEE_FS_SEEK_SET);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	res = ftruncate(fdp->nw_fd, encrypted_data_len);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	/* write encrypted file content to normal world file system */
+	res = write_file(fdp->nw_fd, encrypted_data, encrypted_data_len);
+	if (res != (int)encrypted_data_len) {
+		res = -1;
+		goto exit_free_encrypted_data;
+	}
+	DMSG("%d bytes written", res);
+
+	/* fill in parameters */
+	head.op = TEE_FS_CLOSE;
+	head.fd = fdp->nw_fd;
+
+	res = tee_fs_send_cmd(&head, NULL, 0, TEE_FS_MODE_NONE);
+	if (!res)
+		res = head.res;
+
+exit_free_encrypted_data:
+	free(encrypted_data);
+exit_free_fd:
+	free(fdp->data);
+	free(fdp);
+
+	return res;
+}
+
+static int tee_fs_read(int fd, void *buf, size_t len)
+{
+	int res = -1;
+	size_t remain;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	DMSG("read, fd=%d, buf=%p, len=%d", fd, buf, len);
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (!fdp || !buf)
```

Comment:
Update tee_fs.c accordinly so that we can diff tee_fs.c and tee_enc_fs.c to see only the relevant diffs.